### PR TITLE
Switch endpoints controller to shared informers

### PR DESCRIPTION
**Pull request by ncdc**
cc @bprashanth @thockin @deads2k @sttts @liggitt @smarterclayton  @kubernetes/sig-scalability-pr-reviews

### CODE REVIEW

**Comment by deads2k** [comment]
`Fix controller manager RBAC in local-up-cluster` lgtm

**Comment by deads2k** [comment]
Seems like `Allow controller manager to list/watch Services` should be squashed in here somewhere.

**Comment by deads2k** [comment]
The permission it adds is fine though.

**Comment by ncdc** [comment]
I can squash the policy change with `Switch endpoints controller to shared informers`

File: pkg/controller/endpoint/endpoints_controller.go [hunk 1]
```diff
@@ -338,8 +303,12 @@ func (e *EndpointController) syncService(key string) error {
 		glog.V(4).Infof("Finished syncing service %q endpoints. (%v)", key, time.Now().Sub(startTime))
 	}()
 
-	obj, exists, err := e.serviceStore.Indexer.GetByKey(key)
-	if err != nil || !exists {
+	namespace, name, err := cache.SplitMetaNamespaceKey(key)
+	if err != nil {
+		return err
+	}
+	service, err := e.serviceLister.Services(namespace).Get(name)
```
Comment (deads2k) [inline pkg/controller/endpoint/endpoints_controller.go]: how hard is it to deep copy the map copies that happening?  There's no mutation on `newEndpoints.Labels` as a for instance, but the map is the same as the shared cache for services.

If its non-trivial (it always is in Go), I could live without it.

**Comment by deads2k** [comment]
the controller change lgtm.  I haven't gone back through the resync pull yet.

File: pkg/controller/endpoint/endpoints_controller.go (hunk 1 above)
Comment (ncdc) [inline pkg/controller/endpoint/endpoints_controller.go]: We'd just have to write a for loop to create a new map and copy the KVs in.

**Comment by ncdc** [comment]
@k8s-bot kubemark e2e test this

**Comment by timothysc** [comment]
quick question, why aren't endpoints themselves using informers vs. a hard list + gets.  

/cc @wojtek-t

**Comment by ncdc** [comment]
@timothysc I can convert those too

**Comment by ncdc** [comment]
I think it makes sense to do the endpoint informer change in a follow-up, as updating the test cases is not trivial.

**Comment by timothysc** [comment]
> I think it makes sense to do the endpoint informer change in a follow-up, as updating the test cases is not trivial. 

+1

**Comment by ncdc** [comment]
Rebased on top of the latest in #40759 (no other changes)

**Comment by smarterclayton** [comment]
Rebase and I'll do final review.

**Comment by ncdc** [comment]
Rebased

**Comment by smarterclayton** [comment]
I don't see anything obvious.

**Comment by ncdc** [comment]
Adding the lgtm for Clayton

**Pull request merged**